File: mercury-0.13.1-r2/testenv.m
:- module testenv.
:- interface.

:- import_module io.

:- type testdata --->     
 testdata( % without whitespace between testdata and ( % )
  countOk :: int,
  countFail :: int,
  countLatest :: int
 )
.

:- pred testOk( testdata, testdata).
:- mode testOk(in, out) is det.

:- pred testOk( testdata, testdata, io.state, io.state).
:- mode testOk(in, out, di, uo) is det.

:- pred testFail( testdata, testdata, io.state, io.state).
:- mode testFail(in, out, di, uo) is det.

:- pred checkComplainAndAdjustExpected( testdata, testdata, int, io.state, io.state).
:- mode checkComplainAndAdjustExpected( in, out, in, di, uo) is det.

:- func result( testdata) = string .
:- mode result( in) = out is det.

:- implementation.

:- import_module int.
:- import_module string.

testOk( testdata( OK, F, L), testdata( 1 + OK, F, L)) :- true.
testOk( testdata( OK, F, L), testdata( 1 + OK, F, L), !IO) :- 
 io.write_string( "", !IO),
 true
.
testFail( testdata( OK, F, L), testdata( OK, 1 + F, L), !IO) :- 
 io.write_string( "test failed\n", !IO),
 true
.

checkComplainAndAdjustExpected( testdata( OK, F, L), testdata( OK, F, CountTests), Count, !IO) :- 
 CountTests= OK + F,
 CountTestsLocal= CountTests - L,

 (
  Count = CountTestsLocal -> S = ""
  ;
  S= "Expected Test Count (from last checkpoint to here) doesn't match: countTestsLocal:" ++ string( CountTestsLocal) ++ " count: "++ string( Count) ++ "\n"
 ),
 io.write_string( S, !IO),
 true
.

result( testdata( OK, F, L)) = S :-
 S= "tests: ok: "++ string( OK)++ " fail: "++ string( F)++ " countLatest: "++ string( L)
 ,
 true
.

:- end_module testenv.

File: mono-2.6.7-mcs/TestEnv.cs
namespace TestEnv
{
 class TestEnv
 {

  private int countOk = 0;
  private int countFail = 0;
  private int countLatest = 0;

  public void Test( bool b)
  {
   if( b)
   {
    countOk+= 1;
   }
   else
   {
    countFail+= 1;
    System.Console.WriteLine( "test failed");
    // ggf. stacktrace ausgeben
   }
  }

  public bool CheckComplainAndAdjustExpected( int count)
  {
   int countTests= countOk + countFail;
   int countTestsLocal= countTests - countLatest;

   countLatest= countTests;

   if( count == countTestsLocal)
   {
    return true;
   }
   else
   {
    System.Console.WriteLine( "Expected Test Count (from last checkpoint to here) doesn't match: countTestsLocal:" + countTestsLocal + " count: "+ count);
    // ggf. stacktrace ausgeben
    return false;
   }
  }

  public string Result()
  {
   return "tests: ok: "+ countOk+ " fail: "+ countFail+ " countLatest: "+ countLatest;
  }
 }
}
